Begin implementing WebTransport network interface using quic streams

achristensen07 opened the pull request:
#### f08c4eb963415657f9ad044292a86d4095b23aa1
<pre>
Begin implementing WebTransport network interface using quic streams
<a href="https://bugs.webkit.org/show_bug.cgi?id=278967">https://bugs.webkit.org/show_bug.cgi?id=278967</a>

Reviewed by Matthew Finkel.

My initial implementation to get some bytes on the network used h2, but that requires
a bunch of metadata like headers and verbs and status codes, which isn&apos;t like WebTransport.
This uses quic streams to begin sending and receiving bytes in a format that is closer
to what we will want for WebTransport.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.cpp:
(WebKit::NetworkTransportBidirectionalStream::NetworkTransportBidirectionalStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportReceiveStream.cpp:
(WebKit::NetworkTransportReceiveStream::NetworkTransportReceiveStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportReceiveStream.h:
(WebKit::NetworkTransportReceiveStream::identifier const):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.cpp:
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportBidirectionalStreamCocoa.mm: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.cpp.
(WebKit::NetworkTransportBidirectionalStream::NetworkTransportBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportReceiveStreamCocoa.mm: Added.
(WebKit::NetworkTransportReceiveStream::NetworkTransportReceiveStream):
(WebKit::NetworkTransportReceiveStream::receiveLoop):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSendStreamCocoa.mm: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.cpp.
(WebKit::NetworkTransportSendStream::NetworkTransportSendStream):
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::streamReceiveBytes):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h:
(TestWebKitAPI::WebTransportServer::bytesReceived const): Deleted.
(TestWebKitAPI::WebTransportServer::setBytesReceived): Deleted.
(): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::connectionLoop): Deleted.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.h: Removed.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.mm: Removed.

Canonical link: <a href="https://commits.webkit.org/283097@main">https://commits.webkit.org/283097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ef8321bb586fe69e0d7ff936484abc44a617c2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65172 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44539 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17786 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69196 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15778 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67290 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52322 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16060 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52359 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10916 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68238 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41176 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56417 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32984 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37847 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13788 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14654 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59740 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70901 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9124 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13606 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59690 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/64852 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9156 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56477 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59965 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/ignored-objects, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7542 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1216 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9886 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40351 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41428 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42609 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41172 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportReceiveStream.h:
```diff
@@ -25,19 +25,45 @@
+
+namespace WebKit {
+struct WebTransportStreamIdentifierType;
+using WebTransportStreamIdentifier = LegacyNullableObjectIdentifier<WebTransportStreamIdentifierType>;
```
I'm not a reviewer of course (so I just left this as a comment)--but saw this and had a question/suggestion. 

Feel free to correct me if I'm wrong: I believe we're trying to move away from using `LegacyNullableObjectIdentifier`. Is there a specific reason we need it here? Would it be possible to use `ObjectIdentifier` instead? I was just curious so I thought I'd ask. Thank you!

achristensen07 replied:
You're right, but this is just another declaration of the existing WebTransportStreamIdentifier, which should be changed in a separate PR.

RupinMittal replied:
Got it--that makes sense!

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
```diff
@@ -128,20 +122,12 @@ void NetworkTransportSession::streamReceiveBytes(WebTransportStreamIdentifier id
-    ASSERT(!m_receiveStreams.contains(identifier));
-    m_receiveStreams.set(identifier, makeUniqueRef<NetworkTransportReceiveStream>(*this));
-    send(Messages::WebTransportSession::ReceiveIncomingUnidirectionalStream(identifier));
+    // FIXME: Implement and send Messages::WebTransportSession::ReceiveIncomingUnidirectionalStream.
```
Will this one and `receiveBidirectionalStream` be platform-agnostic? Curious whether the implementations will be in this class, or in cocoa.

achristensen07 replied:
I think it'll be in cocoa or other platforms.

sysrqb commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -28,37 +28,51 @@
-        m_connections.append(connection);
-        connectionLoop(*this, connection);
+    nw_listener_set_new_connection_group_handler(m_listener.get(), ^(nw_connection_group_t connectionGroup) {
+        constexpr uint32_t maximumMessageSize { std::numeric_limits<uint32_t>::max() };
```
Is this a relatively arbitrary limit, or is that specified?

achristensen07 replied:
It requires a parameter and I don't want to say to have a limit for now.  We may revisit.

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -136,10 +151,65 @@
+        if (!completionHandler)
+            return;
+        RefPtr strongThis = weakThis.get();
+        if (!strongThis || !stream)
```
Minor nit: can move `!stream` into the first early return.

achristensen07 replied:
I disagree.  The first early return is only for if completionHandler has already been called.  That will probably be needed for handling network failures after connection establishment.  The second early return is for handling failures before connection establishment.

sysrqb replied:
Okay, that distinction wasn't obvious, but makes sense.

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -46,16 +44,36 @@
 {
-    RetainPtr endpoint = adoptNS(nw_endpoint_create_url(url.string().utf8().data()));
+    // FIXME: Use nw_endpoint_create_url to support all URLs on systems where rdar://135036170 is fixed.
+    bool isIPv4Address = URL::hostIsIPAddress(url.host()) && !url.host().contains(':');
```
More of a general comment: seems like we should provide a variant of `hostIsIPAddress` that returns whether it's an IPv4 or v6 address.

achristensen07 replied:
Not a bad idea.

achristensen07 replied:
It looks like nw_endpoint_create_host accepts more than just IP addresses.  I'll try removing this restriction.

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -46,16 +44,36 @@
-    RetainPtr endpoint = adoptNS(nw_endpoint_create_url(url.string().utf8().data()));
+    // FIXME: Use nw_endpoint_create_url to support all URLs on systems where rdar://135036170 is fixed.
+    bool isIPv4Address = URL::hostIsIPAddress(url.host()) && !url.host().contains(':');
+    if (!url.port() || !isIPv4Address)
```
Is there a reason this requires an explicit port?

achristensen07 replied:
I guess we could use the default port for the protocol if it's not set.  Either way, we're stuck with only IP addresses for now, unfortunately.

sysrqb commented on Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
```diff
@@ -106,6 +106,8 @@ void WebTransportSession::streamReceiveBytes(WebTransportStreamIdentifier identi
     if (auto source = m_readStreamSources.get(identifier))
         source->receiveBytes(bytes, withFin);
+    else
+        ASSERT_NOT_REACHED();
```
Are you planning on adding other error handling for this?

achristensen07 replied:
I don't think I can.  If there's no stream, there's nowhere for the error to go.  It's already been torn down.

sysrqb reviewed (approved):
Looks okay to me, just a couple comments and questions.

achristensen07 committed:
Begin implementing WebTransport network interface using quic streams
https://bugs.webkit.org/show_bug.cgi?id=278967

Reviewed by Matthew Finkel.

My initial implementation to get some bytes on the network used h2, but that requires
a bunch of metadata like headers and verbs and status codes, which isn't like WebTransport.
This uses quic streams to begin sending and receiving bytes in a format that is closer
to what we will want for WebTransport.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.cpp:
(WebKit::NetworkTransportBidirectionalStream::NetworkTransportBidirectionalStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportReceiveStream.cpp:
(WebKit::NetworkTransportReceiveStream::NetworkTransportReceiveStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportReceiveStream.h:
(WebKit::NetworkTransportReceiveStream::identifier const):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.cpp:
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportBidirectionalStreamCocoa.mm: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportBidirectionalStream.cpp.
(WebKit::NetworkTransportBidirectionalStream::NetworkTransportBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportReceiveStreamCocoa.mm: Added.
(WebKit::NetworkTransportReceiveStream::NetworkTransportReceiveStream):
(WebKit::NetworkTransportReceiveStream::receiveLoop):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSendStreamCocoa.mm: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportSendStream.cpp.
(WebKit::NetworkTransportSendStream::NetworkTransportSendStream):
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::streamReceiveBytes):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h:
(TestWebKitAPI::WebTransportServer::bytesReceived const): Deleted.
(TestWebKitAPI::WebTransportServer::setBytesReceived): Deleted.
(): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::connectionLoop): Deleted.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.h: Removed.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.mm: Removed.

Canonical link: https://commits.webkit.org/283097@main
